Adds SyntaxGenerator.ConditionalAccessExpression

bkoelman opened the pull request:
I missed this while writing a language-independent FixProvider.

Pilchie commented:
Sorry this got missed - tagging @dotnet/roslyn-ide and @mattwar for review.

@bkoelman Looks like it has conflicts now, can you resolve and update?

Pilchie commented on src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:
```diff
@@ -3468,6 +3468,11 @@ public override SyntaxNode ExpressionStatement(SyntaxNode expression)
 
+        public override SyntaxNode ConditionalAccessExpression(SyntaxNode expression, SyntaxNode whenNotNull)
+        {
+            return SyntaxFactory.ConditionalAccessExpression(Parenthesize(expression), Parenthesize(whenNotNull));
```
The `whenNotNull` part can't be parenthesized, can it?

Pilchie commented:
Also - can you write at least a single test of the functionality for each language?

CyrusNajmabadi commented:
Agree with kevin on both the comments.  With tests this would LGTM.

bkoelman commented:
Resolved merge conflicts.

@Pilchie @CyrusNajmabadi I've added tests. And updated implementations, but to be honest: it's a bit of a guess... how about this?

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:
```diff
@@ -3531,6 +3531,11 @@ public override SyntaxNode ExpressionStatement(SyntaxNode expression)
 
+        public override SyntaxNode ConditionalAccessExpression(SyntaxNode expression, SyntaxNode whenNotNull)
+        {
+            return SyntaxFactory.ConditionalAccessExpression(Parenthesize(expression), SyntaxFactory.MemberBindingExpression((SimpleNameSyntax)whenNotNull));
```
1. Please wrap very long lines :)
2. No... this would not be ok.  THe problem here is that it assumes the "whenNotNull" will be a simple name.  Really, whenNotNull should be a MemberBindingExpression itself.  If you need to create a MemberBindingExpression from a simple name, then there should be a factory method for that.

bkoelman replied:
1. Yes, I would prefer that too. Is there a maximum length agreed on somewhere?
2. Thanks for explaining. The thing is: the extra factory method that is needed varies for C# (`SyntaxFactory.MemberBindingExpression`) and VB (`SyntaxFactory.SimpleMemberAccessExpression`). How would that factory method be named then, or should it not be in the language-independent interface?

bkoelman replied:
Also, now that I am reusing the existing method `MemberAccessExpression` for extra tests, the built expression gets invalid parenthesis, such as:
```csharp
VerifySyntax<ConditionalAccessExpressionSyntax>(_g.ConditionalAccessExpression(_g.IdentifierName("x"), _g.MemberAccessExpression(_g.MemberBindingExpression(_g.IdentifierName("y")), _g.IdentifierName("z"))), "(x)?(.y).z");
```

where `(x)?(.y).z` is not legal C#. Am I missing something?

CyrusNajmabadi replied:
Yes.  Your code generation logic is wrong :)   As kevin mentioned, you should not be parenthesizing the whenNotNull portion.

CyrusNajmabadi replied:
> Yes, I would prefer that too. Is there a maximum length agreed on somewhere?
Informally: if it's too long for github PRs, it's too long.

CyrusNajmabadi replied:
> The thing is: the extra factory method that is needed varies for C# (SyntaxFactory.MemberBindingExpression) and VB (SyntaxFactory.SimpleMemberAccessExpression)

I think the idea is you would do this:

```f.ConditionalAccess(left, f.MemberBindingExpression(whateverExpression))```

c# would wrap the expression in an actual MemberBindingExpression, VB would just pass the expression along untouched.

bkoelman replied:
That's not what is causing this issue, this happens after removing my
parenthesizing. It's an existing method that's adding them.
Op 21 jun. 2016 02:27 schreef "CyrusNajmabadi" <notifications@github.com>:

> In src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs
> <https://github.com/dotnet/roslyn/pull/11647#discussion_r67790671>:
>
> > @@ -3531,6 +3531,11 @@ public override SyntaxNode ExpressionStatement(SyntaxNode expression)
> >              return SyntaxFactory.ExpressionStatement((ExpressionSyntax)expression);
> >          }
> >
> > +        public override SyntaxNode ConditionalAccessExpression(SyntaxNode expression, SyntaxNode whenNotNull)
> > +        {
> > +            return SyntaxFactory.ConditionalAccessExpression(Parenthesize(expression), SyntaxFactory.MemberBindingExpression((SimpleNameSyntax)whenNotNull));
>
> Yes. Your code generation logic is wrong :) As kevin mentioned, you should
> not be parenthesizing the whenNotNull portion.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/roslyn/pull/11647/files/717e2490e2a35559b6db7f2ff6b05e0f83b9ce52#r67790671>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AJ2JlKF8va4z2BV301bJCpCpbJNlvxitks5qNy_sgaJpZM4IqIKZ>
> .
>

CyrusNajmabadi replied:
if you the make the cahnges discussed, and push to this PR, i can test things out to see what the issue is.

bkoelman replied:
I've pushed my changes so far.

bkoelman replied:
@CyrusNajmabadi did you get a chance to test out the changes in this PR?

CyrusNajmabadi commented on src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs:
```diff
@@ -300,6 +300,7 @@ public void TestConditionalExpressions()
         {
             VerifySyntax<BinaryExpressionSyntax>(_g.CoalesceExpression(_g.IdentifierName("x"), _g.IdentifierName("y")), "(x) ?? (y)");
             VerifySyntax<ConditionalExpressionSyntax>(_g.ConditionalExpression(_g.IdentifierName("x"), _g.IdentifierName("y"), _g.IdentifierName("z")), "(x) ? (y) : (z)");
+            VerifySyntax<ConditionalAccessExpressionSyntax>(_g.ConditionalAccessExpression(_g.IdentifierName("x"), _g.IdentifierName("y")), "(x)?.y");
```
Let's add some more tests as well.  We'd want this to work for things like   a?.b.c  etc.
